man/admin/fsattach.man
'\" Copyright 1989 Regents of the University of California
'\" Permission to use, copy, modify, and distribute this
'\" documentation for any purpose and without fee is hereby
'\" granted, provided that this notice appears in all copies.
'\" The University of California makes no representations about
'\" the suitability of this material for any purpose.  It is
'\" provided "as is" without express or implied warranty.
'\" 
'\" $Header: /sprite/src/admin/fsattach/RCS/fsattach.man,v 1.5 91/01/12 16:49:27 jhh Exp $ SPRITE (Berkeley)
'/" 
.so \*(]ltmac.sprite
.HS FSATTACH cmds
.BS
'\" Note:  do not modify the .SH NAME line immediately below!
.SH NAME
fsattach \- Check disks and export prefixes 
.SH SYNOPSIS
\fBfsattach\fR [\fIoptions\fR]
.SH OPTIONS
.IP "\fB\-m \fIfile\fR" 15
Read mount table from \fIfile\fR.
.IP "\fB\-W\fR" 15
Check disks, but do not write changes.
.IP "\fB\-f\fR" 15
Don't check disks (fastboot).
.IP "\fB\-s\fR" 15
Ignore group information found in mount table and run \fBfscheck\fR 
sequentially.
.IP "\fB\-v\fR" 15
Have both \fBfsattach\fR and \fBfscheck\fR  output verbose information
while running.
.IP "\fB\-n\fR" 15
Don't attach disks or export prefixes.
.IP "\fB\-p\fR" 15
Print out actions but don't execute them.
.IP "\fB\-help\fR" 15
Print a summary of the command-line options and exit.
.IP "\fB\-d \fIdeviceDirectory\fR" 15
Specifies the directory containing the device files.
The default is "/dev".
.IP "\fB\-k\fR" 15
Produce debugging output.
.IP "\fB\-i \fIspriteID\fR" 15
Causes prefixes that this machine will export to be preloaded into the
prefix table. 
This prevents the machine from broadcasting for its own prefixes while
the disks are being checked.
.IP "\fB\-j \fImaxJobs\fR" 15
The default behavior is for \fBfsattach\fR to run as many \fBfscheck\fR
jobs in parallel as it can.
The parallelism is limited by the \fIgroup\fR of each device (see below),
and the \fB\-j\fR flag.
If this flag is used then at most \fImaxJobs\fR \fIfscheck\fRs will
be run at a time.
.IP "\fB\-fscheck \fIfscheckProg\fR" 15
Use this to specify an alternative fscheck program, i.e. when testing
a new fscheck.
.IP "\fB\-c\fR 15
Conditionally check the disks.  If the disks were just checked by \fIfscheck\fR
and we are now rebooting, then we don't have to check again.  If the \fB\-c\fR
flag is present then the disks won't be rechecked. 
.BE

.SH DESCRIPTION
.PP
\fBFsattach\fR is used during system boot to check the disks for 
inconsistencies and
to export prefixes.
The input to \fBfsattach\fR is a mount table, usually in 
\fI/hosts/$HOST/mount\fR.
The location of the mount table is specified by the
\fB\-m\fR option.
\fBFsattach\fR uses the pass information in the mount table to create child
processes that check
the disk partitions using \fBfscheck\fR.
The output from \fBfscheck\fR is appended to a file with the same name as 
the device and the suffix ".fsc".
If \fBfscheck\fR completes normally, 
then the filesystem is mounted and its prefix exported.
.SH MOUNT TABLE 
The mount table contains information on checking the disks and on exporting
prefixes.
Each line in the table consists of a command and its arguments. 
Blank lines and lines that begin with '#' are ignored.
The \fBAttach\fR command is used to check a disk partition and to attach the 
partition to the file system. 
\fBAttach\fR commands have the following format:
.sp
.in 5
\fBAttach\fR \fIprefix\fR \fIdevice\fR \fIgroup\fR \fIscope\fR \fIaccess\fR
.in 0
.LP
\fIPrefix\fR is the prefix under which to mount the partition. 
.LP
The \fIdevice\fR is the
device on which to run fscheck and to attach.
.LP
The \fIgroup\fR parameter is used to determine the parallelism in checking the
disks. \fBFscheck\fR is a very IO constrained application. 
Most of the execution time is spent waiting for the disk. 
If the system has more than one disk the overall run time of \fBfsattach\fR can
be decreased by checking different disks at the same time.
\fIFsattach\fR will run jobs in parallel if possible.
Any partitions that are in the same group  will NOT be checked
in parallel.
A group name is any sequence of characters and is case sensitive.
The group name "root" has special significance.
It is used as the group of the root partition.
This partition is checked before all other partitions, 
and the server may have to reboot if an error is found in this partition.
.LP
The \fIscope\fR can be either local or export. 
Partitions with local scope are attached only for the local machine, 
whereas partitions with export scope are attached and exported under the
given prefix.
The prefix corresponding to the root partition (usually /bootTmp) is
attached during system boot and is not reattached by \fBfsattach\fR.
For this partition the scope parameter does not have any effect.
.LP
The \fIaccess\fR is used to set the access permissions of the prefix. 
If the access is "r" then the prefix is read only. 
If the access is "rw" then the prefix is can be both read and written.
The access parameter has no effect if the partition is the root partition.
.PP
The \fBExport\fR command is used to export a prefix under a different name.
\fBExport\fR commands have the following format:
.sp
.in 5
\fBExport\fR \fIprefix\fR \fIlocalPath\fR 
.in 0
.LP
\fIPrefix\fR is the name of the prefix to be exported.
.LP
\fILocalPath\fR is the name of a path that is to be exported as \fIprefix\fR.
.PP
All other non-comment lines are interpreted as containing arguments to
\fBfscheck\fR.
The format of these lines is:
.sp
.in 5
\fIdeviceName\fR \fI[arguments]\fR
.in 0
.LP
\fIDeviceName\fR is the name of the device for which \fBfscheck\fR is to use the
\fIarguments\fR. A device name of \fBall\fR will cause \fBfscheck\fR to use the 
arguments for all devices. Any number of arguments can follow the device
name, 
as long as they fit on a single line. 
The same device name can appear on any number of lines.
.PP
The following is a sample mount table. The root partition is 
\fI/dev/rsd0a\fR and
it is exported under the prefix \fI/t1\fR. 
\fI/t2\fR is another small partition on
the same disk. 
\fI/b\fR is a large partition on the disk, and \fI/a\fR is a large partition
that is an entire second disk.
.KS
.DS 
#<Command>      <Prefix>        <Device/Path> <Group>   <Scope>       <Access>

Attach          /t1        	/dev/rsd0a     root     export          rw
Attach          /t2             /dev/rsd0b      A       export          rw
Attach          /b              /dev/rsd0g      A       export          rw
Attach          /a              /dev/rsd1c      B       export          rw

Export		/foo		/t2

#<Device>       <Arguments to fscheck>

all             -readBlock 10
/dev/rsd0b      -outputFile foo
.DE
.KE
.SH RETURN CODES
\fBFsattach\fR returns one of the following codes upon exit.
.IP 0
No error.
.IP 1
An error was found in the root partition. Reboot the system. See \fBfscheck\fR
for details.
.IP 2
A hard error occurred. There was some problem in either \fBfsattach\fR, 
\fBfscheck\fR,
or \fBprefix\fR that prevented \fBfsattach\fR from completing normally.
.IP 3 
\fBFscheck\fR returned a soft error indication. See \fBfscheck\fR for details.
.IP 4
An error was found in the root partition, but the number of consecutive
times that \fBfscheck\fR was run and an error found exceeds the limit.
.KS
.SH FILES
.DS L
.nf
/boot/cmds/fscheck			command to check disks
/hosts/$HOST/*.fsc			output from fscheck
/bootTmp/.fscheck.out			temporary output from fscheck
.DE
.KE
.fi
.SH KEYWORDS
fsattach, fscheck, prefix, mount, disk, partition, crash, attach, boot 



src/cmds/fsattach/fsattach.c
/* 
 * fsattach.c --
 *
 *	Boot-time program to attach all disks to a server.  This
 *	uses a mount file that indicates what prefixes/disks to
 *	attach.  This also siphons off some availability/reliability
 *	information that is put in the summary information sector
 *	on each disk.
 *
 * Copyright 1988 Regents of the University of California
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies.  The University of California
 * makes no representations about the suitability of this
 * software for any purpose.  It is provided "as is" without
 * express or implied warranty.
 */

#ifndef lint
static char rcsid[] = "$Header: /sprite/src/admin/fsattach/RCS/fsattach.c,v 1.10 91/01/12 16:48:11 jhh Exp $ SPRITE (Berkeley)";
#endif not lint

#include "fsattach.h"

char *mountFile = "mount";
char *devDir = "/dev/";
char *fscheck = "fscheck";
int verboseLevel = -1;
Boolean nomount = FALSE;
Boolean sequential = FALSE;
Boolean printOnly = FALSE;
Boolean writeDisk = TRUE;
Boolean fastboot = FALSE;
Boolean debug = FALSE;
int	spriteID = 0;
int	maxChildren = -1;
Boolean condCheck = FALSE;

Option optionArray[] = {
    {OPT_STRING,   "d", (char *) &devDir,
	"Device directory."},
    {OPT_STRING,   "fscheck", (char *) &fscheck,
	"fscheck program."},
    {OPT_TRUE,   "f", (char *) &fastboot,
	"Don't check disks."},
    {OPT_INT,   "i", (char *) &spriteID,
	"Preload prefix table with ourself as server of prefixes we export"},
    {OPT_INT,   "j", (char *) &maxChildren,
	"Maximum number of fscheck jobs to run at a time"},
    {OPT_TRUE,   "k", (char *) &debug,
	"Print debugging output."},
    {OPT_STRING,   "m", (char *) &mountFile,
	"File containing disk<=>prefix information."},
    {OPT_TRUE,   "n", (char *) &nomount,
	"No mount."},
    {OPT_TRUE,   "p", (char *) &printOnly,
	"Don't do anything. Just print out actions."},
    {OPT_TRUE,   "s", (char *) &sequential,
	"Ignore group information and run fscheck sequentially."},
    {OPT_TRUE,   "v", (char *) &verbose,
	"Verbose output from fsattach and fscheck."},
    {OPT_FALSE,   "W", (char *) &writeDisk,
	"Don't let fscheck write to the disks."},
    {OPT_TRUE, "c", (char *) &condCheck,
	"Conditionally check the disks (don't re-check)."},

};
int numOptions = Opt_Number(optionArray);

char		*progName;
int		returnCode;

int 		mountTableSize = 30;
int		mountTableSizeIncrement = 5;
MountInfo	*mountTable;
int		groupInfoSize = 10;
int		groupInfoSizeIncrement = 2;
GroupInfo	*groupInfo;
int		numGroups;
char		*tempOutputFile = ".fscheck.out";
int		tempOutputFileSize = 8192;
char		*heapLimitString = "1000000";
Boolean		verbose = FALSE;
Boolean 	reboot = FALSE;




/*
 *----------------------------------------------------------------------
 *
 * main --
 *
 *	Main program for "fsattach":  attach disks at boottime.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

main(argc, argv)
    int argc;
    char *argv[];
{
    int 		mountCount = 0;
    int 		i;
    int 		j;
    ReturnStatus	status;
    int			numChecks;

    argc = Opt_Parse(argc, argv, optionArray, numOptions, 0);

    returnCode = OK;
    progName = argv[0];
    Alloc(mountTable, MountInfo, mountTableSize, "mountTable");
    assert(mountTable != NULL);
    for (i = 0; i < mountTableSize; i++) {
	mountTable[i].status = CHILD_OK;
    }
    Alloc(groupInfo, GroupInfo, groupInfoSize, "groupInfo");
    assert(groupInfo != NULL);
    bzero(groupInfo, sizeof(GroupInfo) * groupInfoSize);
    strcpy(groupInfo[0].name, "root");
    numGroups = 1;
    status = ParseMount(mountFile, &mountCount);
    if (status != SUCCESS) {
	exit(HARDERROR);
    }
    if (spriteID != 0) {
	PreloadPrefixTable(spriteID, mountCount);
    }
    /*
     * We don't want to check the same partition twice. If two entries in
     * the parse table have the same source then set doCheck on one to
     * false. Also count how many partitions have to be done in each pass.
     */
    numChecks = 0;
    for (i = 0; i < mountCount; i++) {
	for (j = i+1; j < mountCount; j++) {
	    if (!strcmp(mountTable[i].source, mountTable[j].source)) {
		mountTable[j].doCheck = FALSE;
	    }
	}
	if (mountTable[i].doCheck == TRUE) {
	    numChecks++;
	}
    }
    if (!fastboot) {
	CacheWriteBack(FALSE);
	CheckDisks(mountCount, numChecks);
	if (!reboot) {
	    CacheWriteBack(TRUE);
	}
    }
    if (reboot) {
	exit(REBOOT);
    }
    if (!nomount) {
	Prefix(mountCount);
    }
    MoveOutput(mountCount);

    if (debug) {
	printf("(1) Exiting with %d.\n", returnCode);
    }
    (void) exit(returnCode);
}

/*
 *----------------------------------------------------------------------
 *
 * CheckDisks --
 *
 *	Check the disks. For each pass fork of an fscheck process for
 *	each partition to be checked.
 *
 * Results:
 *	FAILURE if an error occurred, SUCCESS otherwise.
 *
 * Side effects:
 *	MountTable entries are modified.
 *
 *----------------------------------------------------------------------
 */

ReturnStatus
CheckDisks(mountCount, numChecks)
    int			mountCount;	/* # entries in mount table */
    int			numChecks;	/* # of partitions to check */
{
    int			i;
    ChildInfo		*childInfo;
    int			pass;
    int			doneChildren;
    int			activeChildren;
    union wait 		waitStatus;
    ReturnStatus	status;
    int			childPid;
    MountInfo		*mountPtr;

    Alloc(childInfo, ChildInfo, numChecks, "childInfo");
    assert(childInfo != NULL);
    for (i = 0; i < numChecks; i++) {
	childInfo[i].pid = -1;
    }
    activeChildren = 0;
    doneChildren = 0;
    if (maxChildren < 0) {
	maxChildren = numGroups;
    }
    while (doneChildren < numChecks) {
	while (activeChildren < maxChildren) {
	    status = RunChild(mountCount, numChecks, childInfo, &i);
	    if (status == FAILURE) {
		if (i < 0) {
		    /*
		     * There are no available jobs that can be run.
		     * Wait for some to finish.
		     */
		    assert(activeChildren > 0);
		    break;
		}
		/*
		 * Some sort of error occurred.
		 */
		mountTable[i].doCheck = FALSE;
		returnCode = HARDERROR;
		doneChildren++;
	    } else {
		mountTable[i].status = CHILD_RUNNING;
		groupInfo[mountTable[i].group].running = TRUE;
		activeChildren++;
		if (sequential) {
		    break;
		}
	    }
	}
	if (activeChildren > 0) {
	    if (printOnly) {
		for (i = 0; i < mountCount; i++) {
		    if (mountTable[i].status == CHILD_RUNNING) {
			break;
		    }
		}
		assert(i != numChecks);
		childPid = i;
	    } else {
		childPid = wait(&waitStatus);
	    }
	    activeChildren--;
	    for (i = 0; i < numChecks; i++) {
		if (childInfo[i].pid == childPid) {
		    mountPtr = &(mountTable[childInfo[i].mountIndex]);
		    childInfo[i].pid = -1;
		    break;
		}
	    }
	    if (verbose) {
		printf("Fscheck of %s finished.\n", mountPtr->source);
	    }
	    if (printOnly) {
		doneChildren++;
		mountPtr->status = CHILD_OK;
		mountPtr->doCheck = FALSE;
		groupInfo[mountPtr->group].running = FALSE;
		continue;
	    }
	    if (WIFEXITED(waitStatus) && 
		waitStatus.w_retcode == FSCHECK_OUT_OF_MEMORY) {
		continue;
	    }
	    doneChildren++;
	    mountPtr->doCheck = FALSE;
	    mountPtr->checked = TRUE;
	    groupInfo[mountPtr->group].running = FALSE;
	    if (WIFSIGNALED(waitStatus) || WIFSTOPPED(waitStatus)) {
		(void) fprintf(stderr,"%s did not finish.\n", fscheck);
		returnCode = HARDERROR;
		mountPtr->status = CHILD_FAILURE;
	    } else if (waitStatus.w_retcode == EXEC_FAILED) {
		returnCode = HARDERROR;
		mountPtr->status = CHILD_FAILURE;
	    } else {
		if (debug) {
		    printf("%s returned 0x%x.\n", fscheck,
			(unsigned int) waitStatus.w_retcode);
		    printf("returnCode is %d.\n", returnCode);
		}
		mountPtr->status = CHILD_OK;
		if ((char) waitStatus.w_retcode < 0 ) {
		    PrintFscheckError((char)waitStatus.w_retcode, mountPtr);
		    mountPtr->status = CHILD_FAILURE;
		    returnCode = HARDERROR;
		} else if ((char) waitStatus.w_retcode > 0) {
		    PrintFscheckError((char)waitStatus.w_retcode, mountPtr);
		    if ((char) waitStatus.w_retcode == FSCHECK_REBOOT) {
			reboot = TRUE;
		    }
		    if (returnCode == OK) {
			returnCode = SOFTERROR;
		    }
		} else if (verbose) {
		    PrintFscheckError((char)waitStatus.w_retcode, mountPtr);
		}
	    }
	}
    }
    assert(doneChildren == numChecks);
    assert(activeChildren == 0);
}

/*
 *----------------------------------------------------------------------
 *
 * CacheWriteBack --
 *
 *	Turns cache write-back on and off
 * Results:
 *	None.
 *
 * Side effects:
 *	The cache write-back status is changed.
 *
 *----------------------------------------------------------------------
 */

void
CacheWriteBack(value)
    int		value;
{

    int			newValue;
    ReturnStatus	status;
    int			lockedBlocks;

    if (printOnly && verbose) {
	printf("Setting cache write-back to %d.\n", value);
	return;
    }
    if (!writeDisk) {
	if (verbose) {
	    fprintf(stderr, 
		    "Fscheck not writing disks -- not changing  write-back.\n");
	}
	return;
    }
    newValue = value;
    status = Fs_Command(FS_DISABLE_FLUSH, sizeof(int), (Address) &value);
    if (status != SUCCESS) {
	(void) fprintf(stderr, "Fs_Command (1)  returned %d.\n", status);
	(void) exit(HARDERROR);
    }
    if (verbose) {
	(void) fprintf(stderr, "Cache write-back %s, was %s.\n", 
		(newValue) ? "on" : "off",
	        (value) ? "on" : "off");
    }
    /*
     * If we're turning the write-back off flush what's in the cache already.
     */
    if (newValue == 0) {
	status = Fs_Command(FS_EMPTY_CACHE, sizeof(int), 
		    (Address) &lockedBlocks);
	if (status != SUCCESS) {
	    (void) fprintf(stderr, "Fs_Command (2)  returned %d.\n", status);
	    (void) exit(HARDERROR);
	}
	if (lockedBlocks > 0) {
	    fprintf(stderr, "There are %d locked blocks in the cache ?!\n", 
		lockedBlocks);
	}
    }
}

/*
 *----------------------------------------------------------------------
 *
 * RunChild --
 *
 *	Forks a process to run fscheck.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	A process is forked.
 *
 *----------------------------------------------------------------------
 */

ReturnStatus
RunChild(mountCount, numChecks, childInfo, mountIndexPtr)
    int		mountCount;		/* # entries in mount table */
    int		numChecks;		/* # of jobs to run */
    ChildInfo	*childInfo;		/* info on running children */
    int		*mountIndexPtr;		/* ptr to index into mount table */
{
    int		mountIndex;
    int		i;
    MountInfo	*mountPtr = NULL;
    static char	deviceName[MAX_FIELD_LENGTH];
    static char	partition[MAX_FIELD_LENGTH];
    static char	outputFile[MAX_FIELD_LENGTH];
    int		pid;
    ArgHeader	*argPtr;
    Boolean	defaultOutputFile = TRUE;
    Boolean	defaultRootPart = TRUE;
    Boolean	defaultDir = TRUE;

    /*
     * Find a child to run.
     */
    *mountIndexPtr = -1;
    for (mountIndex = 0; mountIndex < mountCount; mountIndex++) {
	if (mountTable[mountIndex].group == 0 &&
	    mountTable[mountIndex].status == CHILD_RUNNING) {
	    break;
	}
	if ((mountTable[mountIndex].doCheck) &&
	    (groupInfo[mountTable[mountIndex].group].running == FALSE) &&
	    (mountTable[mountIndex].status != CHILD_RUNNING)) {
	    mountPtr = &(mountTable[mountIndex]);
	    break;
	}
    }
    if (mountPtr == NULL) {
	return FAILURE;
    }
    *mountIndexPtr = mountIndex;
    /*
     * Put together the arguments to fscheck.
     */
    (void) strcpy(partition, &(mountPtr->source[strlen(mountPtr->source) -1 ]));
    (void) strcpy(deviceName, mountPtr->source);
    deviceName[strlen(mountPtr->source) - 1] = '\0';
    StartExec(fscheck, fscheck);
    AddExecArgs("-dev", deviceName, NULL);
    AddExecArgs("-part", partition, NULL);
    LIST_FORALL((List_Links *) &mountPtr->argInfo.argList, 
	(List_Links *) argPtr) {
	if (!strcmp("-verbose", argPtr->arg) && verbose) {
	    continue;
	} else if (!strcmp(argPtr->arg, "-write") && !writeDisk) {
	    continue;
	} else if (!strcmp("-dir", argPtr->arg)) {
	    defaultDir = FALSE;
	}
	AddExecArgs(argPtr->arg, NULL);
    }
    DeleteList(&mountPtr->argInfo.argList);
    if (verbose) {
	AddExecArgs("-verbose", NULL);
    }
    if (writeDisk) {
	AddExecArgs("-write", NULL);
    }
    AddExecArgs("-outputFile", tempOutputFile, NULL);
    AddExecArgs("-rawOutput", NULL);
    if (condCheck) {
	AddExecArgs("-cond", "-setCheck", NULL);
    }
    if (defaultDir) {
	AddExecArgs("-dir", devDir, NULL);
    }
    pid = DoExec();
    if (pid < 0) {
	if (verbose) {
	    fprintf(stderr, "Fork of child failed.\n");
	    perror("");
	}
	return FAILURE;
    }
    /*
     * Store info about the running child.
     */
    for (i = 0; i < numChecks; i++) {
	if (childInfo[i].pid == -1) {
	    if (printOnly) {
		childInfo[i].pid = mountIndex;
	    } else {
		childInfo[i].pid = pid;
	    }
	    childInfo[i].mountIndex = mountIndex;
	    break;
	}
    }
    assert(i != numChecks);
    return SUCCESS;
}

/*
 *----------------------------------------------------------------------
 *
 * Prefix --
 *
 *	Adds all devices that were checked correctly to the prefix table.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Entries are added to the system prefix table.
 *
 *----------------------------------------------------------------------
 */

void
Prefix(count)
    int		count;		/* # entries in mount table */
{
    int			i;
    ReturnStatus	status;
    int			flags;
    Fs_TwoPaths		paths;
    char		buffer[128];
    char		*source;
    char		*dest;
    Boolean		device;
    char		*prefix;

    for (i = 0; i < count; i ++) {
	if (mountTable[i].status == CHILD_OK) {
	    flags = 0;
	    if (mountTable[i].readonly) {
		flags |= FS_ATTACH_READ_ONLY;
	    }
	    if (mountTable[i].export == FALSE) {
		flags |= FS_ATTACH_LOCAL;
	    }
	    source = mountTable[i].source;
	    dest = mountTable[i].dest;
	    device = mountTable[i].device;
	    /*
	     * Use Fs_AttachDisk to attach a disk.
	     */
	    if (device == TRUE) {
		printf("Attaching %s%s as %s.\n", devDir, source, dest);
		printf("%s %s.\n", mountTable[i].readonly ? "R" : "RW",
		       mountTable[i].export ? "Export" : "Local");
		if (printOnly) {
		    continue;
		}
		sprintf(buffer, "%s%s", devDir, source);

		status = Fs_AttachDisk(buffer, dest, flags);
		if (status == FS_DOMAIN_UNAVAILABLE) {
		    printf("%s is already attached.\n", buffer);
		    prefix = GetAttachName(buffer);
		    if (prefix != NULL) {
			source = prefix;
			device = FALSE;
		    }
		} else if (status != SUCCESS) {
		    (void) fprintf(stderr, "Attach \"%s\" on \"%s\": %s\n", 
			    buffer, dest, Stat_GetMsg(status));
		}
	    }
	    if (device == FALSE) {
		printf("Exporting %s as %s.\n", source, dest);
		if (printOnly) {
		    continue;
		}
		paths.pathLen1 = strlen(source) +1;
		paths.path1 = source;
		paths.pathLen2 = strlen(dest) +1;
		paths.path2 = dest;
		status = Fs_Command(FS_PREFIX_EXPORT, sizeof(paths), 
			            (Address) &paths);
		if (status != SUCCESS) {
		    fprintf(stderr, 
			    "Couldn't export  \"%s\" as \"%s\": %s\n",
			    source, dest, Stat_GetMsg(status));
		}
	    }
	}
    }
}
